crates/matrix-sdk-ui/src/timeline/mod.rs

```diff
@@ -117,11 +118,17 @@ impl Timeline {
     #[instrument(skip_all, fields(room_id = ?self.room().room_id(), ?options))]
     pub async fn paginate_backwards(&self, mut options: PaginationOptions<'_>) -> Result<()> {
         let mut start_lock = self.start_token.lock().await;
-        if start_lock.is_none()
-            && self.inner.items().await.front().map_or(false, |item| item.is_timeline_start())
-        {
-            warn!("Start of timeline reached, ignoring backwards-pagination request");
-            return Ok(());
+        if start_lock.is_none() {
+            if self.inner.items().await.front().map_or(false, |item| item.is_timeline_start()) {
+                warn!("Start of timeline reached, ignoring backwards-pagination request");
+                return Ok(());
+            }
+
+            if options.wait_for_token {
+                info!("No prev_batch token, waiting");
+                start_lock =
+                    self.start_token_condvar.wait_until(start_lock, |tok| tok.is_some()).await;
```

Comment:
In the case of sliding sync, we know we'll always get a token because the server has to get us the `required_state` when we subscribe. But we can also introduce a timeout if you want. One or two seconds seems like it should be enough, right?

Reply:
I would even go for 3 or 5 secs. Network can be slow. What happens if it times out?